src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/AesGcm.macOS.cs

```diff
@@ -0,0 +1,128 @@
+
+namespace System.Security.Cryptography
+{
+    public sealed partial class AesGcm
+    {
+        // Apple CryptoKit does not support short authentication tags. Since .NET originally supported AES-GCM via
+        // OpenSSL, which does support short tags, we need to continue to support them. If a caller supplies a short
+        // tag we will continue to use OpenSSL if it is available. Otherwise, use CryptoKit.
+        private static readonly bool s_openSslAvailable = Interop.OpenSslNoInit.OpenSslIsAvailable;
+        private const int CryptoKitSupportedTagSizeInBytes = 16;
+
+        private byte[]? _key;
+
+        // CryptoKit added AES.GCM in macOS 10.15, which is our minimum target for macOS. We still may end
+        // up throwing a platform not supported if a caller uses a short authentication tag and OpenSSL is not
+        // available. But recommended use of AES-GCM with a 16-byte tag is supported.
+        public static bool IsSupported => true;
+
+        [MemberNotNull(nameof(_key))]
+        private void ImportKey(ReadOnlySpan<byte> key)
+        {
+            // We should only be calling this in the constructor, so there shouldn't be a previous key.
+            Debug.Assert(_key is null);
+
+            // Pin the array on the POH so that the GC doesn't move it around to allow zeroing to be more effective.
+            _key = GC.AllocateArray<byte>(key.Length, pinned: true);
+            key.CopyTo(_key);
+        }
+
+        private void EncryptCore(
+            ReadOnlySpan<byte> nonce,
+            ReadOnlySpan<byte> plaintext,
+            Span<byte> ciphertext,
+            Span<byte> tag,
+            ReadOnlySpan<byte> associatedData)
+        {
+            CheckDisposed();
+
+            if (tag.Length != CryptoKitSupportedTagSizeInBytes)
+            {
+                using (SafeEvpCipherCtxHandle ctxHandle = CreateOpenSslHandle())
+                {
+                    AesGcmOpenSslCommon.Encrypt(ctxHandle, nonce, plaintext, ciphertext, tag, associatedData);
+                }
+            }
+            else
+            {
+                Interop.AppleCrypto.AesGcmEncrypt(
+                    _key,
+                    nonce,
+                    plaintext,
+                    ciphertext,
+                    tag,
+                    associatedData);
+            }
+        }
+
+        private void DecryptCore(
+            ReadOnlySpan<byte> nonce,
+            ReadOnlySpan<byte> ciphertext,
+            ReadOnlySpan<byte> tag,
+            Span<byte> plaintext,
+            ReadOnlySpan<byte> associatedData)
+        {
+            CheckDisposed();
+
+            if (tag.Length != CryptoKitSupportedTagSizeInBytes)
+            {
+                using (SafeEvpCipherCtxHandle ctxHandle = CreateOpenSslHandle())
+                {
+                    AesGcmOpenSslCommon.Decrypt(ctxHandle, nonce, ciphertext, tag, plaintext, associatedData);
+                }
+            }
```

Comment:
@GrabYourPitchforks How do you feel about the hybrid CryptoKit/OpenSSL approach?  My instinct is to just take it as a breaking change and say that on macOS GCM only supports 16-byte tags... treating it (sort of) like if the OS updated out from under us.  (When OpenSSL raised the minimum RSA from 384 to 512 we updated RSAOpenSsl for the .NET platform version, not the underlying OpenSSL version)

Reply:
@vcsjones, the person you didn't ask, also thinks this is a very reasonable idea.

Reply:
+1 on this. OpenSSL is not part of standard installation on macOS. However, it could be more common on dev machines and inadvertently lure the devs into the trap where they develop something that won't work on most Mac machines.

Reply:
@bartonjs @GrabYourPitchforks how do we want to proceed here? I can address the rest of the feedback depending on what direction we want to take here.

Reply:
Done!